File: tests/snapshot-transport.test.ts
import {describe, expect, test} from "vitest";

import {SnapshotTransport, type SnapshotFile} from "../src/snapshot-transport.ts";
import type {Message, ZulipEvent} from "../src/types.ts";

function fakeMessage(id: number, channel: string, topic: string): Message {
    return {
        id,
        senderId: 1,
        senderFullName: "Test User",
        senderEmail: "test@example.com",
        avatarUrl: "",
        timestamp: 1_700_000_000_000 + id,
        content: `<p>msg ${String(id)}</p>`,
        contentIsHtml: true,
        type: "channel",
        channelName: channel,
        topic,
        reactions: [],
    };
}

function snapshot(messages: Message[]): SnapshotFile {
    return {
        version: 1,
        generatedAt: 1_700_000_000_000,
        server: "https://chat.zulip.org",
        channel: "announce",
        topic: "Zulip updates",
        messages,
    };
}

describe("SnapshotTransport", () => {
    test("serves inline data filtered to scope", async () => {
        const data = snapshot([
            fakeMessage(1, "announce", "Zulip updates"),
            fakeMessage(2, "announce", "other topic"),
            fakeMessage(3, "general", "Zulip updates"),
            fakeMessage(4, "announce", "Zulip updates"),
        ]);
        const transport = new SnapshotTransport({
            url: "unused://",
            scope: {channel: "announce", topic: "Zulip updates"},
            data,
        });
        const events: ZulipEvent[] = [];
        await transport.connect((e) => events.push(e));

        const page = await transport.getMessages({
            channel: "announce",
            topic: "Zulip updates",
        });
        expect(page.messages.map((m) => m.id)).toEqual([1, 4]);
        expect(page.hasMore).toBe(false);
        expect(events[0]).toEqual({type: "connection", status: "connected"});
    });

    test("rejects writes", async () => {
        const transport = new SnapshotTransport({
            url: "unused://",
            scope: {channel: "announce"},
            data: snapshot([]),
        });
        await transport.connect(() => {});
        await expect(
            transport.sendMessage({type: "channel", channel: "announce", content: "hi"}),
        ).rejects.toThrow(/read-only/);
        await expect(transport.addReaction({messageId: 1, emoji: "tada"})).rejects.toThrow(
            /read-only/,
        );
    });

    test("rejects dangerous URL schemes at construction", () => {
        const scope = {channel: "announce", topic: "Zulip updates"};
        expect(() => new SnapshotTransport({url: "javascript:alert(1)", scope})).toThrow();
        expect(() => new SnapshotTransport({url: "data:application/json,{}", scope})).toThrow();
        expect(() => new SnapshotTransport({url: "//evil.tld/x.json", scope})).toThrow(
            /protocol-relative/,
        );
        expect(() => new SnapshotTransport({url: "file:///etc/passwd", scope})).toThrow();
    });

    test("accepts relative and http(s) URLs", () => {
        const scope = {channel: "announce"};
        expect(
            () => new SnapshotTransport({url: "./snapshots/a.json", scope}),
        ).not.toThrow();
        expect(
            () => new SnapshotTransport({url: "https://example.com/a.json", scope}),
        ).not.toThrow();
    });

    test("pagination returns empty on beforeId — snapshots have no backlog", async () => {
        const transport = new SnapshotTransport({
            url: "unused://",
            scope: {channel: "announce", topic: "Zulip updates"},
            data: snapshot([fakeMessage(1, "announce", "Zulip updates")]),
        });
        await transport.connect(() => {});
        const older = await transport.getMessages(
            {channel: "announce"},
            {beforeId: 1},
        );
        expect(older.messages).toEqual([]);
        expect(older.hasMore).toBe(false);
    });
});

File: packages/flutter/test/snapshot_transport_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:zulip_embed/zulip_embed.dart';

Map<String, dynamic> _fixture({
  String channel = 'general',
  String topic = 'welcome',
}) {
  return {
    'version': 1,
    'generatedAt': 1700000000000,
    'server': 'https://chat.zulip.org',
    'channel': channel,
    'topic': topic,
    'messages': [
      {
        'id': 1,
        'senderId': 10,
        'senderFullName': 'Alice',
        'senderEmail': 'a@x',
        'avatarUrl': '',
        'timestamp': 1700000000000,
        'content': '<p>hello</p>',
        'contentIsHtml': true,
        'type': 'channel',
        'channelName': channel,
        'topic': topic,
        'reactions': [],
      },
      {
        'id': 2,
        'senderId': 11,
        'senderFullName': 'Bob',
        'senderEmail': 'b@x',
        'avatarUrl': '',
        'timestamp': 1700000001000,
        'content': 'off-topic',
        'contentIsHtml': false,
        'type': 'channel',
        'channelName': channel,
        'topic': 'other',
        'reactions': [],
      },
    ],
  };
}

void main() {
  group('SnapshotTransport', () {
    test('loads snapshot over http and filters to scope', () async {
      final client = MockClient((req) async {
        expect(req.url.toString(), 'https://example.test/snap.json');
        return http.Response(jsonEncode(_fixture()), 200);
      });
      final transport = SnapshotTransport(
        url: 'https://example.test/snap.json',
        httpClient: client,
      );
      final events = <ZulipEvent>[];
      await transport.connect(
        scope: const ScopeFilter(channel: 'general', topic: 'welcome'),
        onEvent: events.add,
      );
      final messages = await transport.getMessages(
        scope: const ScopeFilter(channel: 'general', topic: 'welcome'),
      );
      expect(messages, hasLength(1));
      expect(messages.first.id, 1);
      expect(events.last, isA<ConnectionEvent>());
      await transport.close();
    });

    test('accepts inline data without hitting the network', () async {
      final transport = SnapshotTransport(
        url: 'unused',
        inlineData: _fixture(),
      );
      await transport.connect(
        scope: const ScopeFilter(channel: 'general'),
        onEvent: (_) {},
      );
      final messages = await transport.getMessages(
        scope: const ScopeFilter(channel: 'general'),
      );
      // No topic filter — both messages survive.
      expect(messages, hasLength(2));
      await transport.close();
    });

    test('rejects sendMessage with a clear error', () async {
      final transport = SnapshotTransport(
        url: 'unused',
        inlineData: _fixture(),
      );
      await transport.connect(
        scope: const ScopeFilter(channel: 'general'),
        onEvent: (_) {},
      );
      await expectLater(
        transport.sendMessage(
          const ChannelSendParams(
            channel: 'general',
            topic: 'general chat',
            content: 'hi',
          ),
        ),
        throwsA(isA<StateError>()),
      );
      await transport.close();
    });

    test('rejects unsafe url schemes at construction', () {
      expect(
        () => SnapshotTransport(url: 'javascript:alert(1)'),
        throwsArgumentError,
      );
      expect(
        () => SnapshotTransport(url: 'file:///etc/passwd'),
        throwsArgumentError,
      );
      expect(
        () => SnapshotTransport(url: '//evil.tld/snap.json'),
        throwsArgumentError,
      );
      expect(
        () => SnapshotTransport(url: ''),
        throwsArgumentError,
      );
    });

    test('rejects unsupported snapshot version', () async {
      final transport = SnapshotTransport(
        url: 'unused',
        inlineData: {
          'version': 2,
          'generatedAt': 0,
          'server': '',
          'channel': 'general',
          'messages': [],
        },
      );
      await expectLater(
        transport.connect(
          scope: const ScopeFilter(channel: 'general'),
          onEvent: (_) {},
        ),
        throwsA(isA<FormatException>()),
      );
    });
  });
}
